doc/specs/#5000 - Process Model 2.0/#5000 - Process Model 2.0.md

```diff
@@ -0,0 +1,1170 @@
+these types that not only enables us to use them in-proc (via the
+implementation), but also enables using these types _across process boundaries_.
+
+Typically, these classes are given a unique GUID for the class. A process that
+wants to implement one of these out-of-proc WinRT objects can register with the
+system to say "I make `MyClass`'s, and their GUID is `{foo}`". These are called
+WinRT _servers_. Then, consumers (_clients_) of that class can ask the system
+"I'd like to instantiate a `{foo}` please", which will cause the server to
+create a new instance of that object (in the server process), and the client
+will be given a WinRT object which can interact with the classes WinRT
+projection.
+
+A server can register the types it produces with `CoRegisterClassObject`, like so:
+
+```c++
+CoRegisterClassObject(MyClassGUID,
+                      winrt::make<MyClassFactory>().get(),
+                      CLSCTX_LOCAL_SERVER,
+                      REGCLS_MULTIPLEUSE,
+                      &dwRegistration);
+
+```
+
+And a client can attempt to get an instance of `MyClass` with
+
+```c++
+auto myClass = create_instance<winrt::MyClass>(MyClassGUID, CLSCTX_LOCAL_SERVER);
+```
+
+We're going to be using that system a little differently here. Instead of using
+a GUID to represent a single _Class_, we're going to use the GUID to uniquely
+identify _content processes_. Each content process will receive a unique GUID on
+creation. It will register as the server for that GUID. Any window process will
+be able to connect to that specific content process strictly by GUID. Because
+each GUID is unique to each content process, any time any client calls
+`create_instance<>(theGuid, ...)`, it will uniquely attempt to connect to the
+content process hosting `theGuid`.
+
+If we wanted to have a second window process connect to the
+_same_ content process as another window, all it needs is the content process's
+GUID.
+
+Now that we have a WinRT object that the content process hosts, we can have the
+window and content process interact using that interface. The next important
+thing to communicate between these processes is the swapchain. The swapchain
+will need to be created by the content process. We also need to be able to
+communicate the `HANDLE` to that swapchain out to the window process. The window process needs to be able to attach that swapchain to the `SwapChainPanel` in the window.
+
+This is tricky, because WinRT does not natively expose a `HANDLE` type. That
+would allow us to easily pass the `HANDLE` between these processes. Instead
+we'll need to manually cast the `HANDLE` to `uint64_t` at the WinRT boundary,
+and cast it back to a `HANDLE` when we want to use it.
+
+Instead, the window will always need to be the one responsible for calling
+`DuplicateHandle`. Fortunately, the `DuplicateHandle` function does allow a
+caller to duplicate from another process into your own process.
+

```

Comment:
idle thoughts -- probably covered in live spec review. I wonder if we should have a MIDL2 interface that lets us use [niksa's recent midl work] somewhere? once we get all the pieces in place.

Reply:
Yea, I think we're going to end up with a split COM/WinRT interface for this. COM to do the niksa thing, WinRT to _not want to blow my brains out doing everything else_